src/gtk/settings.cpp

```diff
@@ -183,40 +188,132 @@ static void notify_gtk_font_name(GObject*, GParamSpec*, void*)
-
-    // 0: No preference, 1: Prefer dark appearance, 2: Prefer light appearance
-    gboolean preferDark = g_variant_get_uint32(value) == 1;
+    {
+        wxLogTrace(TRACE_DARKMODE, "Failed to get GTK settings");
+        return nullptr;
+    }
 
     char* themeName = nullptr;
-    gboolean preferDarkPrev = false;
+    *preferDarkPrev = FALSE;
     g_object_get(settings,
         "gtk-theme-name", &themeName,
-        "gtk-application-prefer-dark-theme", &preferDarkPrev, nullptr);
+        "gtk-application-prefer-dark-theme", preferDarkPrev, nullptr);
 
-    // This is not supposed to happen neither, but don't crash if it does.
+    // This is not supposed to happen, but don't crash if it does.
     if (!themeName)
-        return false;
+    {
+        wxLogTrace(TRACE_DARKMODE, "Failed to get GTK theme name");
+        return nullptr;
+    }
 
-    // We don't need to enable prefer-dark if the theme is already dark
     if (strstr(themeName, "-dark") || strstr(themeName, "-Dark"))
-        preferDark = false;
+    {
+        wxLogTrace(TRACE_DARKMODE, "Force dark mode for theme \"%s\"", themeName);
+        *preferDarkPrev = TRUE;
+    }
+
     g_free(themeName);
 
-    const bool changed = preferDark != preferDarkPrev;
-    if (changed)
+    return settings;
+}
+
+// This function is private to this module and UpdatePreferDark() should
+// normally be used instead to update any appearance-dependent cached settings,
+// but it's enough to call this one on startup, before we have anything cached.
+bool UpdatePreferDarkWithoutRefresh(gboolean preferDark)
+{
+    gboolean preferDarkPrev = FALSE;
+    GtkSettings* const settings = GetPreferDark(&preferDarkPrev);
+    if ( !settings )
+    {
+        // This shouldn't happen, but don't bother doing anything else if it does.
+        return false;
+    }
+
+    if ( preferDark == preferDarkPrev )
+    {
+        // Also don't do anything if the preference didn't change.
+        wxLogTrace(TRACE_DARKMODE, "Dark mode preference didn't change");
+        return false;
+    }
+
+    wxLogTrace(TRACE_DARKMODE, "Turning dark mode preference %s",
+               preferDark ? "on" : "off");
+
+    g_object_set(settings,
+        "gtk-application-prefer-dark-theme", preferDark, nullptr);
+
+    return true;
+}
+
+gboolean ConvertColorSchemeToPreferDark(GVariant* value)
+{
+    const auto n = g_variant_get_uint32(value);
+
+    wxLogTrace(TRACE_DARKMODE, "Color scheme value is %u", n);
+
+    // 0: No preference, 1: Prefer dark appearance, 2: Prefer light appearance
+    return n == 1;
+}
+
+// Global GDBusProxy for org.freedesktop.portal.Settings initialized by
+// wxSystemSettingsModule.
+GDBusProxy* gs_proxyPortalSettings = nullptr;
+
+} // anonymous namespace
+
+// Functions declared in wx/gtk/private/appearance.h and used by wxApp.
+namespace wxGTKImpl
+{
+
+bool GetPreferDarkFromColorScheme(gboolean* preferDark)
+{
+    GVariant* ret = g_dbus_proxy_call_sync(gs_proxyPortalSettings, "Read",
```

Comment:
It seems possible for `gs_proxyPortalSettings` to be null here.

Reply:
Fixed, thanks.